### PR TITLE
[mqtt.generic] separate command parsing from cached value updating

### LINKED ISSUE #12150: [mqtt] REFRESH command doesn't restore item to current value in MQTT
I have an MQTT channel with separate command and state topics. The Item is set to autoupdate=true (the default). The device at the other end may take a long time updating the state topic with the new value after receiving a command (or not update at all, if it's rejected!)

## Expected Behavior
When the device rejects the command, I should be able to send a REFRESH to the item, and it will restore it to the current state in MQTT. But it does not.

## Current Behavior
It just stays forever, making OpenHAB out-of-sync with the device, with no way to fix it besides restarting the MQTT binding (or maybe just the thing?)

## Possible Solution
Make https://github.com/openhab/openhab-addons/blob/78e214651fa5ea715a65d8464df1108f1632290c/bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelState.java#L334 only execute if you have a state topic and a command topic that aren't the same. If they're the same, that line makes sense to immediately update the cache, because the command you sent _is_ the new state of the item in MQTT. But if a device has separate topics, it might ignore the command, and the state should be able to be restored.

## Steps to Reproduce (for Bugs)
1. Set up an MQTT channel with differing (completely unrelated!) command and state topics. Have the state topic start with a retained value of say "77"
1. Link an item to the channel
1. Send a command, such as "78" to the channel
1. Note that in MQTT, the state topic still has a value of "77"
1. Send a REFRESH to the item.
1. It should now restore to 77, but it doesn't change.

## Context
I have a device that's very slow to acknowledge commands, and occasionally drops them, but I still want to use autoupdate to make my UI easy to interact with. I have written a rule so that when a command is sent to the item, after 30s it will send a REFRESH so that OpenHAB will show the actual state of the world. I would expect it to work, but it's not.

## Your Environment
* OpenHAB Version used: 3.2.0
* Operating System: `Linux 5.4.0-89-generic #100-Ubuntu SMP Fri Sep 24 14:50:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux` (Ubuntu)

### PR DESCRIPTION
**Pull request by ccutrer**
fixes #12150

Previously, Value.update would parse the command, _and_ update the cached
value with that command. Which means that when sending a command towards
MQTT (instead of processing an update from MQTT), the cached value was
unintentionally updated. This prevented the REFRESH command from returning
the most recent value received from the device.

Separating the two concerns also makes the test more obvious what they are
testing, and vastly simplified a kludgy workaround that RollershutterValue
was using to be able to process Commands that aren't States.

I've tested this in my local openhab installation to confirm that an MQTT channel with separate state and command topics, where the state topic does not change after sending a command, the item is set to autoupdate, I can now use REFRESH against the item and the prior state is restored.

### CODE REVIEW

**Comment by jochen314** [comment]
Hi, I really like where this is going :-)
I just have one suggestion: Currently the same `parseCommand` is used for messages received from MQTT and for values to be published to MQTT. Can we separate them? Could we have 2 different parse functions, where one calls the other by default?

I remember, that we do have some problems with the percentage value. During command parsing it is not clear, if the devices or OHs value range should be used. If we would have separate functions it would be clear.

**Comment by ccutrer** [comment]
I totally agree on that point - I was super surprised when I realized _incoming_ values from MQTT went through the same parsing function. Sure, I can add a separate name and default to current implementation relatively easily. I think any change of implementation for particular data types probably belong in their own PR.

**Comment by ccutrer** [comment]
@jochen314 : I've updated the PR with your suggestion. do you have any other feedback?

@lolodomo : can you look at this as well?

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelState.java
Comment (jochen314) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelState.java:205]: The variable name does not match any more. 
It is now a parsedMessage, right?

Comment (ccutrer) [reply bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelState.java]: Fixed with latest commit

**Review by jochen314 — APPROVED** [review]
LGTM

**Comment by ccutrer** [comment]
@lolodomo can this be merged now?

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OpenCloseValue.java
Comment (lolodomo) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/OpenCloseValue.java:80]: No need here to test if type is UNDEF?

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/RollershutterValue.java
Comment (lolodomo) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/RollershutterValue.java:103]: No need here to test if type is UNDEF?

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/Value.java
Comment (lolodomo) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/Value.java:100]: No need here to test if type is UNDEF?

**Review by lolodomo — COMMENTED** [review]
Just questions to be sure

**Comment by lolodomo** [comment]
@ccutrer : any feedback on my comments ?

Note that there are now conflicts on few files.

**Comment by ccutrer** [comment]
> @ccutrer : any feedback on my comments ?
> 
> 
> 
> Note that there are now conflicts on few files.

I think there's a good chance your comments need addressed, but I've been too busy to look at them. I'm
hoping to address them in the next couple days. And I'll also address any merge conflicts.

**Comment by lolodomo** [comment]
@ccutrer ; any chance that you finished?

**Comment by ccutrer** [comment]
@lolodomo: I finally got around to this. and no, there is no chance that `type` in those locations could be `UNDEF` (or `NULL`, or `null`). I've updated the prototypes to be `Command` instead of `Type` to ensure this as well. Those methods are called either from `ChannelState.publishValue`, in which case it's already a Command, or `ChannelState.processMessage`, which used `TypeParser.parseCommand` first, taking the declared types each Value allows, and returning if it's not one of those types.

**Comment by ccutrer** [comment]
@jochen314 @lolodomo any more feedback on this?

**Review by lolodomo — CHANGES_REQUESTED** [review]
The changes look now more impacting than in the original PR. I hope you are really sure of these changes as it looks to me as a significative change of behaviour of the MQTT generic binding.

**Comment by lolodomo** [comment]
I will prefer a second review by @openhab/add-ons-maintainers before merging this PR, I have some difficulties to evaluate the risks caused by these changes on all MQTT bindings in openHAB.

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelState.java
Comment (ccutrer) [reply bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelState.java:351]: Correct, and was the entire impetus for this PR. The disambiguation between the direction of values was a happy side effect.

**Comment by ccutrer** [comment]
+1 additional eyes. It is a significant change. But I have been running it at my home for several months, and I use all of the MQTT sub-bindings (manually configured things/channels; homie; home assistant; ESP Milight Hub). I'm very confident that the only end-user visible change is that if you send a command to an MQTT channel, and the device does not update a corresponding state topic, and then you send REFRESH to the channel, the cached value will not have changed from the command. If you never send REFRESH to the channel, you'll never notice that the cached value wasn't changed. If you're concerned about this, I could see having an additional configuration option to say "assume the published values are accepted, and update internal cache". However, this seems contradictory to me with the auto update framework and state predictions. Maybe a better approach would be a way to say "for this MQTT device, make the AutoUpdatePolicy RECOMMEND rather than DEFAULT." If either of those get recommended, I'd rather implement them as a separate PR though - this one is already large enough.

**Comment by lsiepel** [comment]
Would be nice if this can be merged for the next milestone so we can get some testing milage. I did some basic tests (not using all mqtt sub bindings though) and did not have any issue to report.

**Comment by antroids** [comment]
> @antroids : any chance you'd be willing to look at this?

LGTM since it was tested and didn't broke something.

As for me, I would like to move State out of Value, rename Values to ValueHandlers or MqttTypes and make them immutable. These parse methods could be something like reducers: Command reduceIncoming(Outgoing)MqttMessage(State previousState, Command message). 
This could be another refactoring PR, for sure.

**Comment by lolodomo** [comment]
Are you are several to be confident with these changes, I will merge them.
I have to take a look to additional changes applied after my last review.

**Comment by antroids** [comment]
> I've tested this in my local openhab installation to confirm that an MQTT channel with separate state and command topics, where the state topic does not change after sending a command, the item is set to autoupdate, I can now use REFRESH against the item and the prior state is restored.

Trying to figure out all possible flows.
What if we have Percentage (or any depending on the internal state) channel with the only command topic? 
I don't think that it will work properly after the change.
Maybe we should consider command only channels as optimistic and update the value for them.

**Comment by ccutrer** [comment]
He's referring to IncreaseDecreaseType essentially not working for a channel without a state topic, since we no longer keep track of the last command sent to it. I would argue that it never _should_ have worked. What does it mean to ask a channel to increase if it doesn't keep track of its value? Alternately, we should enable (optionally) passing such commands directly through to MQTT, which is one of the follow up PRs I've been waiting for this one to be merged before working on.

**Comment by antroids** [comment]
> > What if we have Percentage (or any depending on the internal state) channel with the only command topic?
> > I don't think that it will work properly after the change.
> > Maybe we should consider command only channels as optimistic and update the value for them.
> 
> Just before I push the merge button, I read that. Does it mean that I should not merge ?

I think It can be merged. 
It rather addon design issue than error in this particular PR.

**Pull request merged**
